control/setup_control.go

```diff
@@ -0,0 +1,399 @@
+		cancelCtx, cancelFunc := context.WithCancel(context.Background())
+		if err := pidLoop.TunePIDLoop(cancelCtx, cancelFunc); err != nil {
+			return nil, err
+		}
+	}
+
+	if options.NeedsSingleAutoTuning {
+		if err := pidLoop.StartControlLoop(); err != nil {
+			return nil, err
+		}
+	}
+
+	return pidLoop, nil
+}
+
+// TunePIDLoop runs the auto-tuning process for a PID control loop.
+func (p *PIDLoop) TunePIDLoop(ctx context.Context, cancelFunc context.CancelFunc) error {
+	var errs error
+	p.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer func() {
+			cancelFunc()
+			if p.ControlLoop != nil {
+				p.ControlLoop.Stop()
+				p.ControlLoop = nil
+			}
+		}()
+		defer p.activeBackgroundWorkers.Done()
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		// switch sum to depend on the setpoint if position control
+		if p.Options.PositionControlUsingTrapz {
+			p.ControlConf.Blocks[sumIndex].DependsOn[0] = p.BlockNames["constant"][0]
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			if err := p.ControlLoop.MonitorTuning(ctx); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+		}
+		if p.Options.SensorFeedbackVelocityControl {
+			// to tune linear PID values, angular PI values must be non-zero
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"] = 0.5
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"] = 0.5
+			p.logger.Info("tuning linear PID")
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			if err := p.ControlLoop.MonitorTuning(ctx); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			p.ControlLoop.Stop()
+			p.ControlLoop = nil
+
+			// to tune angular PID values, linear PI values must be non-zero
+			p.ControlConf.Blocks[linearPIDIndex].Attribute["kP"] = 0.5
+			p.ControlConf.Blocks[linearPIDIndex].Attribute["kI"] = 0.5
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"] = 0.0
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"] = 0.0
+			p.logger.Info("tuning angular PID")
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			if err := p.ControlLoop.MonitorTuning(ctx); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+		}
+		if p.Options.UseCustomConfig {
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+		}
+	})
+	return errs
+}
+
+func (p *PIDLoop) createControlLoopConfig(pidVals []PIDConfig, componentName string) {
+	// create basic control config
+	if p.Options.ControllableType != "" {
+		controllableType = p.Options.ControllableType
+	}
+	p.basicControlConfig(componentName, pidVals[0], controllableType)
+
+	// add position control
+	if p.Options.PositionControlUsingTrapz {
+		p.addPositionControl()
+	}
+
+	// add sensor feedback velocity control
+	if p.Options.SensorFeedbackVelocityControl {
+		p.addSensorFeedbackVelocityControl(pidVals[1])
+	}
```

Comment:
do we need to validate that these are not being used together?

Reply:
yeah, it was not clear to me if the position and velocity control flags could be used together. On a similar note same with the `CompleteCustomConfig` flag

Reply:
yeah I'll add a check. as for `CompleteCustomConfig`, I was treating that as we shouldn't do anything to it, so we just start the control loop. I can move that check to the top and return so there's no conflicts